### PR TITLE
more verification for vsc packages

### PR DESCRIPTION
**Pull request by stdweird**
* check if vsc namespace is shared with setuptools
* force existence of `README.md`
* generate long_description based on `README.md`
* provide `VSCImportTest` class to test importing packages, modules and scripts
* add all files in `bin/` as `scripts` by default
* strict enforcement of `LICENSE`
* determine `name` if not defined in target from project
 * incl `url` and `download_url` without template
* verify the license header
 * incl hacky tool to add the correct one

### CODE REVIEW

**Comment by stdweird** [comment]
sharing the namespace with setuptools makes it an install requirement for all vsc packages, but looks like it will allow `zip_safe`

File: lib/vsc/install/shared_setup.py [hunk 1]
```diff
@@ -138,6 +139,29 @@
...
+    res = {}
+    offset = len(REPO_LIB_DIR.split(os.path.sep))
+    for root, _, files in os.walk(REPO_LIB_DIR):
+        package='.'.join(root.split(os.path.sep)[offset:])
+        if '__init__.py' in files:
+            if package == 'vsc' or package.startswith('vsc.'):
+                init = open(os.path.join(root, '__init__.py')).read()
+                reg = re.search(r'^import\s+pkg_resources\npkg_resources.declare_namespace\(__name__\)$', init, re.M)
+                if not reg:
+                    raise Exception('vsc namespace packages do not allow non-shared namespace in dir %s. Fix with pkg_resources.declare_namespace' % root)
```
Comment (boegel) [inline lib/vsc/install/shared_setup.py]: line too long?

File: lib/vsc/install/shared_setup.py [hunk 2]
```diff
@@ -138,6 +139,29 @@
...
+def files_in_packages():
+    """
+    Gather all __init__ files provided by the lib/ subdir
+        filenames are relative to the REPO_BASE_DIR
+    Return dict with key the package and value all files in the package directory
+    """
+    res = {}
+    offset = len(REPO_LIB_DIR.split(os.path.sep))
+    for root, _, files in os.walk(REPO_LIB_DIR):
+        package='.'.join(root.split(os.path.sep)[offset:])
```
Comment (boegel) [inline lib/vsc/install/shared_setup.py]: spaces around `=`

File: lib/vsc/install/shared_setup.py [hunk 3]
```diff
@@ -138,6 +139,29 @@
...
+        if '__init__.py' in files:
+            if package == 'vsc' or package.startswith('vsc.'):
+                init = open(os.path.join(root, '__init__.py')).read()
+                reg = re.search(r'^import\s+pkg_resources\npkg_resources.declare_namespace\(__name__\)$', init, re.M)
+                if not reg:
+                    raise Exception('vsc namespace packages do not allow non-shared namespace in dir %s. Fix with pkg_resources.declare_namespace' % root)
+
+            res[package] = [os.path.relpath(os.path.join(root, f),start=REPO_BASE_DIR) for f in files]
```
Comment (boegel) [inline lib/vsc/install/shared_setup.py]: space after `,`

File: lib/vsc/install/shared_setup.py [hunk 4]
```diff
@@ -138,6 +139,29 @@
...
+    """
+    res = {}
+    offset = len(REPO_LIB_DIR.split(os.path.sep))
+    for root, _, files in os.walk(REPO_LIB_DIR):
+        package='.'.join(root.split(os.path.sep)[offset:])
+        if '__init__.py' in files:
+            if package == 'vsc' or package.startswith('vsc.'):
+                init = open(os.path.join(root, '__init__.py')).read()
+                reg = re.search(r'^import\s+pkg_resources\npkg_resources.declare_namespace\(__name__\)$', init, re.M)
+                if not reg:
```
Comment (boegel) [inline lib/vsc/install/shared_setup.py]: `reg` is not reused, so use `if not re.search():` ?

**Comment by boegel** [comment]
Looks OK, I guess.

How extensively did you test the `pkg_resources` stuff?

**Comment by stdweird** [comment]
what do you mean? not enough to turn on zipped eggs just yet, i first need to rebuild my python dev environment so i can test zipped eggs properly

**Comment by boegel** [comment]
Well, if it works as expected with multiple VSC packages around, and with a mix of zipped eggs on non-zipped installs, etc.

**Comment by boegel** [comment]
BTW, I still feel we should fallback to `extend_path` if `pkg_resources` is not available...

File: lib/vsc/__init__.py [hunk 5]
```diff
@@ -1,37 +1,3 @@
...
-so explicitly declare this is also the vsc namespace
-
-@author: Jens Timmerman (Ghent University)
-"""
-from pkgutil import extend_path
-
-# we're not the only ones in this namespace
-__path__ = extend_path(__path__, __name__)  #@ReservedAssignment
+# allow other packages to extend this namespace, zip safe setuptools style
+import pkg_resources
```
Comment (boegel) [inline lib/vsc/__init__.py]: fallback to `pkgutil.extend_path` in case `pkg_resources` is not available (backward compatibility, etc.)?

Comment (stdweird) [inline lib/vsc/__init__.py]: what do you mean? we require setuptools.

Comment (boegel) [inline lib/vsc/__init__.py]: for vsc-install, yes, but not in other places?

Comment (stdweird) [inline lib/vsc/__init__.py]: you mean in like in almost all our python repos were we assume it is available?

**Comment by stdweird** [comment]
tested on vsc-base too

File: lib/vsc/install/shared_setup.py [hunk 6]
```diff
@@ -623,20 +791,118 @@ def sanitize(name):
...
+        raise Exception('UNKONWN LICENSE %s provided. Should be fixed or added to vsc-install' % license)
+
+    log.info("Found license name %s and classifier %s" , lic_short, data[1])
+    return lic_short, data[1]
+
+
+def parse_target(target, urltemplate=None):
     """
     Add some fields
+        get name / ur/ download url from project
```
Comment (JensTimmerman) [inline lib/vsc/install/shared_setup.py]: ur -> url

File: lib/vsc/install/shared_setup.py [hunk 7]
```diff
@@ -623,20 +791,118 @@ def sanitize(name):
...
+        # sets name / url and download_url
+        target.update(get_name_url())
+
+    # prepare classifiers
+    classifiers = new_target.setdefault('classifiers', [])
+
+    if urltemplate:
+        new_target['url'] = urltemplate % target
+        if 'github' in urltemplate:
+            new_target['download_url'] = "%s/tarball/master" % new_target['url']
```
Comment (JensTimmerman) [inline lib/vsc/install/shared_setup.py]: master will surely be wrong, try `%s/archive/%(name)s-V%(version)s.tar.gz`
and make sure the release script does this release on github and pipy

**Comment by stdweird** [comment]
@JensTimmerman remarks addressed

**Pull request merged**